File: elixir/secret-handshake/lib/secret_handshake.ex
defmodule SecretHandshake do
  @doc """
  Determine the actions of a secret handshake based on the binary
  representation of the given `code`.

  If the following bits are set, include the corresponding action in your list
  of commands, in order from lowest to highest.

  1 = wink
  10 = double blink
  100 = close your eyes
  1000 = jump

  10000 = Reverse the order of the operations in the secret handshake
  """
  @spec commands(code :: integer) :: list(String.t())
  def commands(code) do
    code
    |> Integer.digits(2)
    |> Enum.reverse
    |> build(0, [])
  end


  defp build([], _, acc), do: acc |> Enum.reverse

  defp build([h | t], 0, acc) when h == 1, do: build(t, 1, ["wink" | acc])
  defp build([h | t], 0, acc) when h == 0, do: build(t, 1, acc)

  defp build([h | t], 1, acc) when h == 1, do: build(t, 2, ["double blink" | acc])
  defp build([h | t], 1, acc) when h == 0, do: build(t, 2, acc)

  defp build([h | t], 2, acc) when h == 1, do: build(t, 3, ["close your eyes" | acc])
  defp build([h | t], 2, acc) when h == 0, do: build(t, 3, acc)

  defp build([h | t], 3, acc) when h == 1, do: build(t, 4, ["jump" | acc])
  defp build([h | t], 3, acc) when h == 0, do: build(t, 4, acc)

  defp build([h | _], 4, acc) when h == 1, do: acc
  defp build([h | _], 4, acc) when h == 0, do: acc |> Enum.reverse

end

File: c/secret-handshake/secret_handshake.c
#include "secret_handshake.h"
#include <math.h>
#include <stdlib.h>

typedef enum {
  Wink = 1,
  DoubleBlink = 2,
  CloseYourEyes = 4,
  Jump = 8,
  Reverse = 16,
} command_t;

const char **commands(size_t number) {
  int n = log2(number);
  char **code = calloc(n >= 0 ? n + 1 : 1, sizeof(char *));
  int index = 0;
  for (int i = 0; i < n + 1; ++i) {
    command_t c = pow(2, i);
    if ((number & c) == 0) {
      continue;
    }
    switch (c) {
    case Wink:
      code[index++] = "wink";
      break;
    case DoubleBlink:
      code[index++] = "double blink";
      break;
    case CloseYourEyes:
      code[index++] = "close your eyes";
      break;
    case Jump:
      code[index++] = "jump";
      break;
    case Reverse:
      for (int i = 0; i < index / 2; ++i) {
        char *tmp = code[index - i - 1];
        code[index - i - 1] = code[i];
        code[i] = tmp;
      }
    }
  }
  return (const char **)code;
}
